packages/jaeger-ui/src/model/ddg/transformTracesToPaths.tsx

```diff
@@ -35,22 +38,27 @@ function transformTracesToPaths(
           spanMap.set(span.spanID, span);
           return !span.hasChildren;
         })
+
         .forEach(leaf => {
-          const spans = spanAncestorIds(leaf).map(id => {
+          const ancestors = spanAncestorIds(leaf).reverse();
+          ancestors.push(leaf.spanID);
+          const spans = ancestors.reduce((reducedSpans: Span[], id: string): Span[] => {
             const span = spanMap.get(id);
             if (!span) throw new Error(`Ancestor spanID ${id} not found in trace ${traceID}`);
-            return span;
-          });
-          spans.reverse();
-          spans.push(leaf);
-
-          const path: TDdgPayloadEntry[] = spans
-            .filter(span => span.tags.find(({ key, value }) => key === 'span.kind' && value === 'server'))
-            .map(({ processID, operationName: operation }) => ({
-              service: data.processes[processID].serviceName,
-              operation,
-            }));
-
+            if (reducedSpans.length === 0) {
+              reducedSpans.push(span);
+            } else if (
+              reducedSpans[reducedSpans.length - 1].processID !== span.processID ||
+              isKindServer(span)
+            ) {
+              reducedSpans.push(span);
+            }
+            return reducedSpans;
+          }, []);
+          const path: TDdgPayloadEntry[] = spans.map(({ processID, operationName: operation }) => ({
+            service: data.processes[processID].serviceName,
+            operation,
+          }));
```

Comment:
I find this algorithm quite hard to follow, not to mention that it seems to do a lot of redundant work by constantly recalculating the paths from each leaf upwards over and over. Have we tried implementing it as a classic BFS on a tree?

Reply:
We can do that, but one think I would like to avoid is using recursion, it will lead to bad performance (see https://github.com/jaegertracing/jaeger-ui/pull/326).

Can we do that in a separate PR?

Reply:
ok

Reply:
I created another version of this algorithm using a depth breadth search tree algorithm, I'll update this PR soon.

Reply:
Updated the algorithm, I haven't reused the tree , I can do it in another PR if it's required.